src/glomap/estimators/rotation_averaging_impl.cc

```diff
@@ -0,0 +1,803 @@
+    }
+  }
+
+  // Compute current frame rotations for cam_from_rig averaging.
+  std::unordered_map<frame_t, Eigen::Matrix3d> frame_rotations;
+  for (const auto& [frame_id, frame_idx] : frame_id_to_param_idx_) {
+    if (!has_gravity(frame_id)) {
+      frame_rotations[frame_id] = colmap::AngleAxisToRotationMatrix(
+          rotation_estimated_.segment(frame_idx, 3));
+    } else {
+      frame_rotations[frame_id] =
+          colmap::RotationFromYAxisAngle(rotation_estimated_[frame_idx]);
+    }
+  }
+
+  // Update the global rotations for cam_from_rig cameras.
+  // Note: the update is non-trivial, and we need to average the rotations from
+  // all the frames.
+  for (const auto& [camera_id, camera_idx] : camera_id_to_param_idx_) {
+    const Eigen::Matrix3d R_ori = colmap::AngleAxisToRotationMatrix(
+        rotation_estimated_.segment(camera_idx, 3));
+    const Eigen::Matrix3d R_update =
+        colmap::AngleAxisToRotationMatrix(-step.segment(camera_idx, 3));
+
+    std::vector<Eigen::Quaterniond> rig_rotations;
+    for (const frame_t frame_id : camera_to_frame_ids_[camera_id]) {
+      const Eigen::Matrix3d& R = frame_rotations[frame_id];
+      rig_rotations.push_back(
+          Eigen::Quaterniond(R_ori * R * R_update * R.transpose()));
+    }
+
+    // Average the rotations for the rig.
+    const Eigen::Quaterniond R_ave = colmap::AverageQuaternions(
+        rig_rotations, std::vector<double>(rig_rotations.size(), 1));
+    rotation_estimated_.segment(camera_idx, 3) =
+        colmap::RotationMatrixToAngleAxis(R_ave.toRotationMatrix());
+  }
+}
+
+double RotationAveragingProblem::AverageStepSize(
+    const Eigen::VectorXd& step) const {
+  double total_update = 0;
+  for (const auto& [frame_id, frame_idx] : frame_id_to_param_idx_) {
+    const auto it = frame_to_pose_prior_.find(frame_id);
+    const bool has_gravity = options_.use_gravity &&
+                             it != frame_to_pose_prior_.end() &&
+                             it->second->HasGravity();
+    if (has_gravity) {
+      total_update += std::abs(step[frame_idx]);
+    } else {
+      total_update += step.segment(frame_idx, 3).norm();
+    }
+  }
+  return total_update / frame_id_to_param_idx_.size();
+}
+
+void RotationAveragingProblem::ApplyResultsToReconstruction(
+    colmap::Reconstruction& reconstruction) {
+  for (const auto& [frame_id, frame_idx] : frame_id_to_param_idx_) {
+    const auto it = frame_to_pose_prior_.find(frame_id);
+    const bool has_gravity = options_.use_gravity &&
+                             it != frame_to_pose_prior_.end() &&
+                             it->second->HasGravity();
+
+    if (has_gravity) {
+      reconstruction.Frame(frame_id).SetRigFromWorld(Rigid3d(
+          Eigen::Quaterniond(
+              colmap::GravityAlignedRotation(it->second->gravity) *
+              colmap::RotationFromYAxisAngle(rotation_estimated_[frame_idx])),
+          Eigen::Vector3d::Zero()));
```

Comment:
Should we set translation to NaN here or copy the existing value through, if the frame had a pose before?